code/game/objects/items/weapons/tools/mods/_upgrades.dm

```diff
@@ -89,6 +89,14 @@
 
 
 /obj/item/weapon/tool_upgrade/proc/can_apply(var/obj/item/weapon/tool/T, var/mob/user)
+	if (isrobot(T))
+		var/mob/living/silicon/robot/R = T
+		if (R.module_active && istool(R.module_active))
+			try_apply(R.module_active,user)
+		else
+			user << SPAN_WARNING("[src] needs to be wielding a compatible tool.")
```

Comment:
I reckon this will be too fiddly in practise. Instead, allow the user to choose from a list of the robot's compatible tools. Loop through them all, run can_apply on each and present a list of those that return true